# nf-core/scdownstream: Output

:::warning
This section is still the nf-core template output documentation. It will be updated to reflect the output of the nf-core/scdownstream pipeline soon.
:::

## Introduction

This document describes the output produced by the pipeline. Most of the plots are taken from the MultiQC report, which summarises results at the end of the pipeline.

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

<!-- TODO nf-core: Write this documentation describing your workflow's output -->
Reading this you likely have already experienced first hand that single-cell sequencing data is big, exceeding the amount of RAM available to most computers. Nonetheless, algorithms may demand expression values of a particular set of genes for a particular set of cells at any time.
The [hiearchical data format 5](https://en.wikipedia.org/wiki/Hierarchical_Data_Format) (hdf5) thus provides the data in a structured manner with indices such that computational demands are mnimized to retrieve individual records. And much like a traditional file system, or like an excel sheet with many named tabs, different pieces of information can be integrated into a single consistent ..AnnData object that is effectively mapped to a file on the hard drive.

Various steps in the pipeline produce outputs that need to be added to .obs or obsm etc. If the pipeline inputs are really large (10s of GB), always storing the whole AnnData object with the new info and in the end extracting only the relevant info from all objects will use a lot of disk space. To prevent this, this pipeline only stores the delta in each step and finally aggregates all deltas onto a single anndata object.

Each output is put into a directory named like the prefix (until first underline) of the process creating it.
This is still the default of nf-core, but can and should be changed, but has so far not been relevant for functionality development.

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:

- [FastQC](#fastqc) - Raw read QC
- [MultiQC](#multiqc) - Aggregate report describing results and QC from the whole pipeline
- [Pipeline information](#pipeline-information) - Report metrics generated during the workflow execution

### FastQC

<details markdown="1">
<summary>Output files</summary>

- `fastqc/`
  - `*_fastqc.html`: FastQC report containing quality metrics.
  - `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.

</details>

[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences. For further reading and documentation see the [FastQC help pages](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).

![MultiQC - FastQC sequence counts plot](images/mqc_fastqc_counts.png)

![MultiQC - FastQC mean quality scores plot](images/mqc_fastqc_quality.png)

![MultiQC - FastQC adapter content plot](images/mqc_fastqc_adapter.png)

:::note
The FastQC plots displayed in the MultiQC report shows _untrimmed_ reads. They may contain adapter sequence and potentially regions with low quality.
:::

### MultiQC

<details markdown="1">
<summary>Output files</summary>

- `multiqc/`
  - `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
  - `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
  - `multiqc_plots/`: directory containing static images from the report in various formats.

</details>

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

Results generated by MultiQC collate pipeline QC from supported tools e.g. FastQC. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see <http://multiqc.info>.

## Results interpretation


A bit more technical background: When concatenating the anndata objects performing an outer join, the missing values will be filled with 0s instead of NaNs. This is because 0 values use nearly no memory in the sparse representation. Filling them with NaNs will make the memory blow up.

The reports generated within MultiQC by itself are not yet providing the ultimate insights to biologically interpret the findings.  Granting the reports some further time to evolve, we suggest to employ [CELLxGENE])(https://cellxgene.cziscience.com/) for now to help with an integration of public data and the interactive generation of publication-ready figures.  If so required, the following files are generated that can be read by statistic suites to select genes/cells of interest:

<details markdown="1">
<summary>Meant for other tools / own developments</summary>
| File                                          | What it represents |
|-----------------------------------------------|--------------------|
|./adata/merged.h5ad                            | Describes all genes, missing data described as N/As. |
|./adata/merged\_inner.h5ad                     | Describes genes present in all cells. |
|./adata/merged\_outer.h5ad                     | Describes all genes, filling missing data up with 0s. |
</details>

For the inner and outer merged objects:
When concatenating the various input datasets there are two ways of handling the genes: Either performing an inner or an outer join of the genes present in all datasets - that's what you will find here. Biologists often want to see all their favourite genes in the final object, so the final result will contain the outer join. For integration, only the inner join will be used, as missing genes often lead to a clustering that's more based on the presence of absence of genes, than on actual biological differences. Because of this, the integration is performed based on the intersection.

<details markdown="1">
<summary>Output for immediate inspection</summary>
</details>
| File                                          | What it represents |
|-----------------------------------------------|--------------------|
|./adata/merged\_sample\_genes.png              |                    |
</details>

The files below are deltas as generated by individual tools and also included in the final output. Used in isolation, these are likely of little to no use beyond debugging.

<details markdown="1">
<summary>Output files that contribute to the final h5ad file. The [Pickle (pkl)](https://en.wikipedia.org/wiki/Serialization#Python) files are typically serialized data sheets.</summary>
| File                                          | What it represents |
|-----------------------------------------------|--------------------|
|./adata/merged.rds                             |                    |
|./adata/merged\_mqc.json                       |                    |
|./adata/merged\_metadata.csv                   |                    |
|./adata/_samplename_.rds                       |                    |
|./adata/_samplename_\_unfiltered\_unified.h5ad |                    |
|./adata/_samplename_.h5ad                      |                    |
|./celda/_samplename_\_decontx.h5ad             |                    |
|./scvitools/X\_scvi\_scvi.pkl                  |                    |
|./scvitools/scvi\_scvi\_model/model.pt         |                    |
|./scvitools/scvi\_scvi.h5ad                    |                    |
|./scanpy/scvi\_umap.h5ad                       |                    |
|./scanpy/_samplename_\_scrublet.pkl            |                    |
|./scanpy/_samplename_\_scrublet.pkl            |                    |
|./scanpy/_samplename_\_filtered.h5ad           |                    |
|./scanpy/scvi\_neighbors.h5ad                  |                    |
|./scanpy/X\_scvi\_umap.pkl                     |                    |
|./scanpy/scvi-0.5\_leiden.h5ad                 |                    |
|./scanpy/scvi-1.0\_leiden.h5ad                 |                    |
</details>

It is in our very interest to further extend the current more QC-associated reports towards a set of routine figures that would then for instance be expected in the supplement of a publication. Also, for automated high-throughput analyses revisiting archived experiments this pipeline shall be of interest to apply. This has not yet been implemented but [https://github.com/nf-core/scdownstream/issues/2](this early issue on GitHub) gives instructions and invites to extend our current reports.

### Pipeline information

<details markdown="1">
<summary>Output files</summary>

- `pipeline_info/`
  - Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  - Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
  - Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.
  - Parameters used by the pipeline run: `params.json`.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
